### PR TITLE
Feature/cubic interpolation 359

### PR DESCRIPTION
**Pull request by lelemmen**

### CODE REVIEW

File: flik/linesearch/interpolations.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+"""Interpolations for step length selection."""
+import numpy as np
+
+
+def cubic_interpolation(var, func, grad, direction, interval_start, interval_end, alpha_i_, alpha_i):
+    """
+    A cubic interpolation algorithm for functions where the gradient is not expensive to calculate, according to eq.
+    (3.59) in "Numerical Optimization" by Nocedal and Wright.
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: The first line in the docstring should not extend beyond one line. You can move the reference to the References section in the docstring.

File: flik/linesearch/interpolations.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    Parameters
+    ----------
+    var: np.array((N,1), dtype=float)
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: It's not a two dimensional array
`np.ndarray((N,), dtype=float)`

File: flik/linesearch/interpolations.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+"""Interpolations for step length selection."""
+import numpy as np
+
+
+def cubic_interpolation(var, func, grad, direction, interval_start, interval_end, alpha_i_, alpha_i):
+    """
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: Should be a raw string b/c of the latex stuff

File: flik/linesearch/interpolations.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    Raises
+    ------
+        TypeError
+            according to the function check_input
+        ValueError
+            according to the function check_input
+            if interval_start and interval_end are not such that
+                0 < interval_start < interval_end <= 1
+            if alpha_1 and alpha_2 are not such that
+                interval_start < alpha_i_ < alpha_i < interval_end
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: I think these all can be indented once to the left

File: flik/linesearch/interpolations.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    alpha_i: float
+        The i-th step length estimate
+
+    Raises
+    ------
+        TypeError
+            according to the function check_input
+        ValueError
+            according to the function check_input
+            if interval_start and interval_end are not such that
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: I think you can just say
`If 0 < interval_start < interval_end <= 1`

File: flik/linesearch/interpolations.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    -------
+    alpha: float
+        The minimizer :math:`\alpha` of :math:`f(x + \alpha p)`, where alpha is in the provided interval
+        [interval_start, interval_end]
+    """
+
+    # Check the input arguments
+    # TODO: Check the parameters according to the function check_input
+    if isinstance(var, float):
+        var = np.array(var)
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: `np.array(var, dtype=float, ndmin=1)`

File: flik/linesearch/interpolations.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        Function :math:`f(x)`
+    grad: callable
+        Gradient of the function :math:`\nabla f(x)`
+    direction: np.ndarray((N,1)) for N variables
+        Direction vector for next step
+    interval_start: float
+        The starting point of the interval containing desirable step lengths
+    interval_end: float
+        The ending point of the interval containing desirable step lengths
+    alpha_i_: float
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: I'm not too sure about the name of this varaible

File: flik/linesearch/interpolations.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        """The derivative of the previously defined function phi."""
+        return grad(var + alpha * direction).dot(direction)
+
+
+    print(var)
+    print(direction)
+    print(alpha_i_, phi(alpha_i_), phi_(alpha_i_))
+    print(alpha_i, phi(alpha_i), phi_(alpha_i))
+    print(interval_start, phi(interval_start))
+    print(interval_end, phi(interval_end))
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: I think we can delete all of the print statements (we'll do a proper logging later on I think)

File: flik/linesearch/interpolations.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    #   Check for 0 < interval_start < interval_end <= 1
+    if not (0.0 < interval_start < interval_end <= 1.0):
+        raise ValueError("interval_start and interval_end should be according to 0 < interval_start < interval_end <= 1")
+
+    #   Check for interval_start < alpha_1 < alpha_2 < interval_end
+    if not (interval_start < alpha_i_ < alpha_i < interval_end):
+        raise ValueError("alpha_i_ and alpha_i should be according interval_start < alpha_i_ < alpha_i < interval_end")
+
+    # Define the helper functions phi and its derivative
+    def phi(alpha):
```
Comment (kimt33) [inline flik/linesearch/interpolations.py]: docstring

File: flik/linesearch/test/test_interpolations.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
+"""Tests for flik.linesearch.interpolations."""
+
+import numpy as np
+from nose.tools import assert_raises
+
+
```
Comment (kimt33) [inline flik/linesearch/test/test_interpolations.py]: one too many newlines

File: flik/linesearch/test/test_interpolations.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    #
+    # alpha_1 = 0.3
+    # alpha_2 = 0.9
+    # assert_raises(ValueError, cubic_interpolation, var, func, grad, direction, interval_start, interval_end, alpha_1,
+    #               alpha_2)
+
+
+def test_cubic_interpolation():
+    """Test the cubic interpolation algorithm with a manual calculation."""
+
```
Comment (kimt33) [inline flik/linesearch/test/test_interpolations.py]: one too many newline

File: flik/linesearch/test/test_interpolations.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Let's pick the following function: f(x) = x.x
+    def func(x):
+        return x.dot(x)
+
+    def grad(x):
+        return 2 * x
+
```
Comment (kimt33) [inline flik/linesearch/test/test_interpolations.py]: one too many newline (I think)

File: flik/linesearch/test/test_interpolations.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    #
+    # alpha_1 = 0.1
+    # alpha_2 = 0.7
+    # assert_raises(ValueError, cubic_interpolation, var, func, grad, direction, interval_start, interval_end, alpha_1,
+    #               alpha_2)
+    #
+    # alpha_1 = 0.3
+    # alpha_2 = 0.9
+    # assert_raises(ValueError, cubic_interpolation, var, func, grad, direction, interval_start, interval_end, alpha_1,
+    #               alpha_2)
```
Comment (kimt33) [inline flik/linesearch/test/test_interpolations.py]: do these tests fail?

**Review by kimt33 — CHANGES_REQUESTED** [review]
Just some small changes. I think Travis will get most of these when you push again (you need to submit it again to get noticed by Travis).

File: flik/linesearch/interpolations.py [hunk 14]
```diff
@@ -93,19 +93,19 @@ def phi_(alpha):
...
-    alpha_next = alpha_current - \
-                 (alpha_current - alpha_previous) * (phi_(alpha_current) + d_2 - d_1) / \
-                 (phi_(alpha_current) - phi_(alpha_previous) + 2 * d_2)
+    d_2 = np.sign(alpha_current - alpha_previous) * np.sqrt(d_1 ** 2 - phi_(alpha_previous) *
+                                                            phi_(alpha_current))
 
+    alpha_next = alpha_current - (alpha_current - alpha_previous) * (phi_(alpha_current) + d_2 -
+                                                                     d_1) / (phi_(alpha_current) -
+                                                                             phi_(alpha_previous) +
+                                                                             2 * d_2)
```
Comment (lelemmen) [inline flik/linesearch/interpolations.py]: I really don't agree on enforcing "E127 continuation line over-indented for visual indent" on this one. The formulas are completely illegible.

File: flik/linesearch/test/test_interpolations.py (hunk 12 above)
Comment (lelemmen) [reply flik/linesearch/test/test_interpolations.py]: pycodestyle doesn't complain about this one, though.

**Comment by lelemmen** [comment]
At this point Travis only fails because of coverage in flik/linesearch/input.py.

**Review by kimt33 — APPROVED** [review]

**Pull request merged**